### PR TITLE
Ability to modify response headers (mainly cache related headers)

### PR DESCRIPTION
**Pull request by silverbackdan**
The cache headers should be able to be modified and then not overwritten by the `AddHeadersListener`. Also added annotation configuration for cache header expiry times so each resource can have a different expiry time if required.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/api-platform/core/issues/2207
| License       | MIT
| Doc PR        | (not created yet)

### CODE REVIEW

File: src/HttpCache/EventListener/AddHeadersListener.php [hunk 1]
```diff
@@ -30,16 +31,21 @@
     private $sharedMaxAge;
     private $vary;
     private $public;
+    private $resourceMetadataFactory;
 
-    public function __construct(bool $etag = false, int $maxAge = null, int $sharedMaxAge = null, array $vary = null, bool $public = null)
+    public function __construct(ResourceMetadataFactoryInterface $resourceMetadataFactory, bool $etag = false, int $maxAge = null, int $sharedMaxAge = null, array $vary = null, bool $public = null)
```
Comment (antograssiot) [inline src/HttpCache/EventListener/AddHeadersListener.php]: the argument must added in the last position and have a default value for BC

File: src/HttpCache/EventListener/AddHeadersListener.php [hunk 2]
```diff
@@ -30,16 +31,21 @@
...
+        $this->resourceMetadataFactory = $resourceMetadataFactory;
         $this->etag = $etag;
         $this->maxAge = $maxAge;
         $this->sharedMaxAge = $sharedMaxAge;
         $this->vary = $vary;
         $this->public = $public;
     }
 
+    /**
+     * @throws \ApiPlatform\Core\Exception\ResourceClassNotFoundException
```
Comment (antograssiot) [inline src/HttpCache/EventListener/AddHeadersListener.php]: I think that `@throws` tags are not propagated to parent function in the core usually

File: src/HttpCache/EventListener/AddHeadersListener.php [hunk 3]
```diff
@@ -53,23 +59,29 @@ public function onKernelResponse(FilterResponseEvent $event)
             return;
         }
 
-        if ($this->etag) {
+        $resourceCacheHeaders = [];
+        if ($this->resourceMetadataFactory && $attributes = RequestAttributesExtractor::extractAttributes($request)) {
+            $resourceMetadata = $this->resourceMetadataFactory->create($attributes['resource_class']);
+            $resourceCacheHeaders = $resourceMetadata->getAttribute('cache_headers');
```
Comment (antograssiot) [inline src/HttpCache/EventListener/AddHeadersListener.php]: maybe `$resourceMetadata->getOperationAttribute($attributes, 'cache_headers', [], true)`

File: src/HttpCache/EventListener/AddHeadersListener.php [hunk 4]
```diff
@@ -53,23 +59,29 @@ public function onKernelResponse(FilterResponseEvent $event)
...
+        }
+
+        if (!$response->getEtag() && $this->etag) {
             $response->setEtag(md5($response->getContent()));
         }
 
-        if (null !== $this->maxAge) {
-            $response->setMaxAge($this->maxAge);
+        if (!$response->headers->hasCacheControlDirective('max-age') && (($useMeta = isset($resourceCacheHeaders['max_age'])) || null !== $this->maxAge)) {
```
Comment (antograssiot) [inline src/HttpCache/EventListener/AddHeadersListener.php]: maybe : 

```php
if (null !== ($maxAge = $resourceCacheHeaders['max_age'] ?? $this->maxAge) && !$response->headers->hasCacheControlDirective('max-age')) {
    $response->setMaxAge($maxAge);
}
```

File: src/HttpCache/EventListener/AddHeadersListener.php [hunk 5]
```diff
@@ -53,23 +59,29 @@ public function onKernelResponse(FilterResponseEvent $event)
...
         }
 
-        if ($this->etag) {
+        $resourceCacheHeaders = [];
+        if ($this->resourceMetadataFactory && $attributes = RequestAttributesExtractor::extractAttributes($request)) {
+            $resourceMetadata = $this->resourceMetadataFactory->create($attributes['resource_class']);
+            $resourceCacheHeaders = $resourceMetadata->getAttribute('cache_headers');
+        }
+
+        if (!$response->getEtag() && $this->etag) {
```
Comment (antograssiot) [inline src/HttpCache/EventListener/AddHeadersListener.php]: execution cost of the second part is probably lower

File: src/HttpCache/EventListener/AddHeadersListener.php [hunk 6]
```diff
@@ -53,23 +59,29 @@ public function onKernelResponse(FilterResponseEvent $event)
...
+            $response->setMaxAge($useMeta ? $resourceCacheHeaders['max_age'] : $this->maxAge);
         }
 
         if (null !== $this->vary) {
             $response->setVary(array_diff($this->vary, $response->getVary()), false);
         }
 
-        if (null !== $this->sharedMaxAge) {
-            $response->setSharedMaxAge($this->sharedMaxAge);
+        if (!$response->headers->hasCacheControlDirective('s-maxage') && (($useMeta = isset($resourceCacheHeaders['shared_max_age'])) || null !== $this->sharedMaxAge)) {
```
Comment (antograssiot) [inline src/HttpCache/EventListener/AddHeadersListener.php]: same syntax as above can be applied

File: src/HttpCache/EventListener/AddHeadersListener.php [hunk 7]
```diff
@@ -53,23 +59,29 @@ public function onKernelResponse(FilterResponseEvent $event)
...
+            $response->setSharedMaxAge($useMeta ? $resourceCacheHeaders['shared_max_age'] : $this->sharedMaxAge);
         }
 
-        if (null !== $this->public) {
+        if (!$response->headers->hasCacheControlDirective('public') && !$response->headers->hasCacheControlDirective('private') && null !== $this->public) {
```
Comment (antograssiot) [inline src/HttpCache/EventListener/AddHeadersListener.php]: if conditions can be switched too

**Comment by dunglas** [comment]
👍 on my side when @antograssiot's comments will be fixed! Thanks to both of you!

File: src/Annotation/ApiResource.php [hunk 8]
```diff
@@ -256,6 +256,13 @@
      */
     private $sunset;
 
+    /**
+     * @see https://github.com/Haehnchen/idea-php-annotation-plugin/issues/112
+     *
+     * @var array
+     */
+    private $cacheHeaders;
```
Comment (antograssiot) [inline src/Annotation/ApiResource.php]: trully sorry but I forgot to mention that you also need to add ` @Attribute("cacheHeaders", type="array")` at the top of the file (in the alphabetical order) :disappointed:

**Comment by silverbackdan** [comment]
Is it worth re-running behat-coverage? It seems the behat tests all worked fine on other checkers but failed on that one.

File: src/Annotation/ApiResource.php (hunk 8 above)
Comment (silverbackdan) [reply src/Annotation/ApiResource.php]: You had to comment on so many changes, bound to miss one - thanks for helping me improve my skills.

File: src/Annotation/ApiResource.php [hunk 9]
```diff
@@ -53,7 +53,8 @@
  *     @Attribute("subresourceOperations", type="array"),
  *     @Attribute("sunset", type="string"),
  *     @Attribute("swaggerContext", type="array"),
- *     @Attribute("validationGroups", type="mixed")
+ *     @Attribute("validationGroups", type="mixed"),
+ *     @Attribute("cacheHeaders", type="array")
```
Comment (antograssiot) [inline src/Annotation/ApiResource.php]: that'not what I meant by in alphabetical order :stuck_out_tongue:

Comment (silverbackdan) [reply src/Annotation/ApiResource.php]: I'm so dozy, sorry to waste your time like that. I must learn to take time and read comments properly and stop pushing before ready as well. One minute I'll push the change.

**Comment by dunglas** [comment]
Thanks @silverbackdan!

**Pull request merged**